clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/AWSServerGroupNameResolverSpec.groovy

```diff
@@ -0,0 +1,136 @@
+    def takenSlots = resolver.getTakenSlots('application-stack-details')
+
+    then:
+    1 * amazonClusterProvider.getCluster('application', account, 'application-stack-details') >> {
+      // should only include server groups in the target region
+      new SimpleCluster(type: 'aws', serverGroups: [
+          SimpleServerGroup.sG('application-stack-details-v000', 0, region),
+          SimpleServerGroup.sG('application-stack-details-v001', 0, 'us-east-1')
+      ])
+    }
+    1 * googleClusterProvider.getCluster('application', account, 'application-stack-details') >> {
+      // wrong cluster type, should not be included in taken slots
+      new SimpleCluster(type: 'google', serverGroups: [
+        SimpleServerGroup.sG('application-stack-details-v001', 0, region)
+      ])
+    }
+    0 * _
+
+    takenSlots == [
+        new AbstractServerGroupNameResolver.TakenSlot('application-stack-details-v000', 0, new Date(0))
+    ]
+  }
+
+  void "should ensure that next server group does not already exist (and increment as necessary until new server group name is found)"() {
+    given:
+    def clusterProviders = [amazonClusterProvider]
+    def resolver = new AWSServerGroupNameResolver(account, region, asgService, clusterProviders)
+
+    and:
+    def clusterName = 'application-stack-details'
+
+    when:
+    def nextServerGroupName = resolver.resolveNextServerGroupName('application', 'stack', 'details', false)
+
+    then:
+    1 * amazonClusterProvider.getCluster('application', account, clusterName) >> {
+      new SimpleCluster(type: 'aws', serverGroups: [
+        SimpleServerGroup.sG("${clusterName}-v999", 0, region)
+      ])
+    }
+    1 * asgService.getAutoScalingGroup("${clusterName}-v000") >> { null }
+    0 * _
+
+    nextServerGroupName == "${clusterName}-v000".toString()
+
+    when:
+    nextServerGroupName = resolver.resolveNextServerGroupName('application', 'stack', 'details', false)
+
+    then:
+    1 * amazonClusterProvider.getCluster('application', account, clusterName) >> {
+      new SimpleCluster(type: 'aws', serverGroups: [
+        SimpleServerGroup.sG('application-stack-details-v999', 0, region)
+      ])
+    }
+    (0..500).each {
+      1 * asgService.getAutoScalingGroup(String.format("${clusterName}-v%03d", it)) >> { new AutoScalingGroup() }
+    }
+    1 * asgService.getAutoScalingGroup("${clusterName}-v501") >> { null }
+    0 * _
+
+    nextServerGroupName == "${clusterName}-v501".toString()
+  }
+
+  void "should raise IllegalArgumentException if unable to determine next server group"() {
+    given:
+    def clusterProviders = [amazonClusterProvider]
+    def resolver = new AWSServerGroupNameResolver(account, region, asgService, clusterProviders)
+
+    and:
+    def clusterName = 'application-stack-details'
+
+    when:
+    resolver.resolveNextServerGroupName('application', 'stack', 'details', false)
+
+    then:
+    1 * amazonClusterProvider.getCluster('application', account, clusterName) >> {
```

Comment:
This is likely never to happen in reality ... but if someone does happen to have 1000 server groups then we need to fail.